### PR TITLE
Feature/alarms

### PR DESCRIPTION
**Pull request by svranesevic**
ActuatorStatus:
Renamed getRef to getReference

Alarm:
Model

Persistence/InMemoryPersistence:
Added support for Alarm

OutboundMessageFactory/JsonSingleOutboundMessageFactory:
Make OutboundMessage from List<Alarm>

Wolk:
Alarms can be added via addAlarm method
Warn when ActuationHandler/ActuatorStatusProvider are unassigned, but used
Fix formatting

Persistece:
Alarm storing policy

Gradle:
Use MQTT uber-jar with (partially) fixed reconnect mechanism

Wolk:
Fix formatting

Wolk:

MqttFactory:
Tweak reconnect parameters

### CODE REVIEW

File: src/main/java/com/wolkabout/wolk/Wolk.java [hunk 1]
```diff
@@ -295,6 +327,11 @@ public void execute() {
      * @param actuatorReference reference of the actuator
      */
     public void publishActuatorStatus(final String actuatorReference) {
+        if (actuatorStatusProvider == null) {
+            LOG.error("Can not publish actuator status for reference " + actuatorReference + ". No ActuatorStatusProvider assigned to Wolk");
```
Comment (nmaksimovic) [inline src/main/java/com/wolkabout/wolk/Wolk.java]: Throw IllegalStateException... See actuationHandler code above...

File: src/main/java/com/wolkabout/wolk/InMemoryPersistence.java [hunk 2]
```diff
@@ -33,6 +34,14 @@
         return readings.get(key);
     }
 
+    private List<Alarm> getOrCreateAlarmsByKey(String key) {
```
Comment (nmaksimovic) [inline src/main/java/com/wolkabout/wolk/InMemoryPersistence.java]: Sounds so side-effectively wrong :)

File: src/main/java/com/wolkabout/wolk/InMemoryPersistence.java [hunk 3]
```diff
@@ -72,6 +81,44 @@ public void removeReadings(String key, int count) {
...
+    @Override
+    public boolean putAlarm(String key, Alarm alarm) {
+        getOrCreateAlarmsByKey(key).add(alarm);
+        return true;
+    }
+
+    @Override
+    public List<Alarm> getAlarms(String key, int count) {
+        if (!readings.containsKey(key)) {
+            return new ArrayList<>();
```
Comment (nmaksimovic) [inline src/main/java/com/wolkabout/wolk/InMemoryPersistence.java]: Maybe Collections.emptyList() ? Returns empty immutable list.

File: src/main/java/com/wolkabout/wolk/InMemoryPersistence.java [hunk 4]
```diff
@@ -72,6 +81,44 @@ public void removeReadings(String key, int count) {
...
+        final List<Alarm> alarmsByKey = getOrCreateAlarmsByKey(key);
+        alarms.put(key, alarmsByKey.subList(Math.min(alarmsByKey.size(), count), alarmsByKey.size()));
+    }
+
+    @Override
+    public List<String> getAlarmsKeys() {
+        List<String> keys = new ArrayList<>();
+        for (final String key : alarms.keySet()) {
+            if (!alarms.get(key).isEmpty()) {
+                keys.add(key);
```
Comment (nmaksimovic) [inline src/main/java/com/wolkabout/wolk/InMemoryPersistence.java]: keys.addAll(readings.keySet()) ?

File: src/main/java/com/wolkabout/wolk/Wolk.java [hunk 5]
```diff
@@ -139,18 +139,12 @@ public void run() {
...
-        if (!futureConnection.isConnected()) {
-            return false;
-        }
-
         try {
             futureConnection.publish(topic, payload.getBytes(), QoS.EXACTLY_ONCE, false).await(30, TimeUnit.SECONDS);
-            return true;
         } catch (Exception e) {
-            return false;
+            throw new TimeoutException("Publish timeout.");
```
Comment (nmaksimovic) [inline src/main/java/com/wolkabout/wolk/Wolk.java]: Is this checked exception? Is system expected to recover from this?

File: src/main/java/com/wolkabout/wolk/InMemoryPersistence.java [hunk 6]
```diff
@@ -72,6 +81,44 @@ public void removeReadings(String key, int count) {
...
+    }
+
+    @Override
+    public void removeAlarms(String key, int count) {
+        if (!alarms.containsKey(key)) {
+            return;
+        }
+
+        final List<Alarm> alarmsByKey = getOrCreateAlarmsByKey(key);
+        alarms.put(key, alarmsByKey.subList(Math.min(alarmsByKey.size(), count), alarmsByKey.size()));
```
Comment (nmaksimovic) [inline src/main/java/com/wolkabout/wolk/InMemoryPersistence.java]: O.o ?

File: src/main/java/com/wolkabout/wolk/InMemoryPersistence.java (hunk 4 above)
Comment (svranesevic) [reply src/main/java/com/wolkabout/wolk/InMemoryPersistence.java]: That would give us all keys, we want keys with non-empty associated list.

File: src/main/java/com/wolkabout/wolk/Wolk.java (hunk 5 above)
Comment (svranesevic) [reply src/main/java/com/wolkabout/wolk/Wolk.java]: It is checked exception, and system recovers from it

**Pull request merged**